drivers/serial.c

```diff
@@ -137,3 +138,24 @@ int serial_write(io_port_t port, const char *buf, size_t len) {
 
     return rc;
 }
+
+void uart_input_enable(void) {
+    int unmasked_irqs = inb(PIC1_PORT_DATA);
+    int unmask_com = COM1_IRQ_UNMASK & COM2_IRQ_UNMASK & unmasked_irqs;
+
+    outb(PIC1_PORT_DATA, unmask_com); /* unmask com irqs */
+}
+
+void uart_handler(io_port_t ports[2]) {
+    io_port_t port;
+    unsigned int i, status, data_avail;
+    outb(PIC1_PORT_CMD, PIC_EOI);
```

Comment:
Isn't this too early?

Reply:
I've seen this on osdev; I believe the speed of the processor here makes it possible to reset the device, which by the time there is another interrupt, this handler has finished running. I can move it at the end if you prefer though

Reply:
No particular preference, but please test what is better.